octopus_deploy/datadog_checks/octopus_deploy/check.py

```diff
@@ -73,10 +74,31 @@ def _process_endpoint(self, endpoint, params=None, report_service_check=False):
                 self.warning("Failed to access endpoint: %s: %s", endpoint, e)
                 return {}
 
+    def _process_paginated_endpoint(self, endpoint, params=None, report_service_check=False):
+        skip = 0
+        take = self.config.paginated_limit
+        num_pages = 1
+        num_pages_seen = 0
+        all_items = []
+        params = {} if params is None else params
+        while num_pages_seen < num_pages:
+            params['skip'] = skip
+            params['take'] = take
+
+            response_json = self._process_endpoint(endpoint, params=params, report_service_check=report_service_check)
+            if response_json == {}:
+                return response_json
+            items = response_json.get("Items")
+            num_pages_seen += 1
+            num_pages = response_json.get("NumberOfPages", num_pages)
```

Comment:
If num_pages is constantly changing with every iteration of the while loop, I'm not sure what the purpose of the while loop condition is. For example, if an iteration of the loop, num_pages was found to be 5, and num_page_seen were 10 (which is then incremented to 11), then 11 pages would be seen in total, even though num_pages is 5? I'm not sure why we would exit the while loop then.

At a first glance, what I'm thinking is that since it's the same endpoint, with the same "take" value, this value should be the same no matter what iteration of the loop it is in (as total number of items, which is constant / take value, which is constant = number of pages, which therefore should be constant). If that's true, then num_pages wouldn't need to be constantly redefined in the loop.

Reply:
I decided to fetch `NumberOfPages` every loop since although the number of pages is usually the same, it could potentially change if more items are created while we are processing the endpoint. Most of the time it will stay the same, though. An alternative approach is to just loop until we get an empty list of items.